### PR TITLE
Fix swapped indices of jewelry and nirnhoned shields

### PR DESCRIPTION
**Pull request by Sc4rlite**
With Gold Road, the indices of rings and necklaces for the crafting bench changed.

Here's what still works with the game update:

- Checkmarks for learned traits are still shown correctly

Here's what the game update broke:

- Tooltips have wrong information (e.g. a ring might show as not yet researched when it's actually the necklace that's missing)
- Wrong items are marked in the CraftStore grid as being available in inventory (e.g. it might show an exemplary ring from your inventory when hovering over the missing necklace trait)
- No jewelry items can be researched via the CraftStore interface anymore whatsoever (via middle mouse button), even when there is a suitable item and even when the trait is missing for both ring and necklace (e.g. it tries to use a ring item to research a necklace trait and fails with "No free research slot or item inaccessible!" error)

According to my tests, the change in line 3192 fixed all three issues. Unless there's any additional bugs that I didn't test, the change should fix all jewelry again.

The remaining changes are merely to keep all mentions of rings and necklaces in the code consistent with the new order, but aren't actually changing any functionality.

### CODE REVIEW

**Comment by Zeratoxx** [comment]
Hmm.. your fix @Sc4rlite is kinda difficult to accept I would say.. 

First of: Yes, your code fixes the logic;  
But: The saved vars wont getting cleaned up!

## Before

First, I show the current behavior of the current official release of CraftStore.

> ## Current official release
> 
> image
> 
> ## Ring column where necklace is getting displayed
> 
> image
> 
> ## Necklace column where ring is getting displayed
> 
> image

So there you can clearly see the issue with the swapped ID's by ZOS, all good. All as expected.

---

## Now

I manually copied your changes into my local CraftStore code, now it's like this:

> ## State after I manually copied your changes into my local CraftStore code
> 
> image
> 
> ## Ring column where ring AND necklace is getting displayed
> 
> image image
> 
> ## Necklace column where ring AND necklace is getting displayed
> 
> image 
image

So now, the entries for stored researchable items are doubled. 2/4 times correct and the other 2/4 times wrong.

---

## Investigations

In my investigations, I tried to delete the false entries in the file `CraftStoreFixedAndImproved.lua` in the `SavedVariables` folder at first.  
This didn't help unfortunately - the reason is getting explained later. After a `/reloadui` with my changes to the file in the folder `SavedVariables`, the result is the same as shown in the pictures from when I manually applied your code changes manually.  
Also, my adjustements to the file in the folder `SavedVariables` got reverted by the `/reloadui`.

Now.. there are also SavedVariables in the RAM of the game.. And this is where the issue is to be found. My changes to the file in the folder `SavedVariables` wont get acknowledged and will get overwritten by the stored values of the variables in the game RAM.  

---

## Further investigations with Torchbug

I was able to alter the live varibles in the RAM of the game with the debug addon Torchbug.

> ## Before altering vars with Torchbug 
> 
> image image
> 
> ## After altering vars with Torchbug
> 
> image image
> 
> ## Result
> 
> image
> As it should be

## Conclusion

Soo, if the live vars in the RAM of the game wont get fixed, the issue just broadens and will get more difficult to get fixed in following patches.  
We need to actively fix the live vars of the game with the loading of the UI. I assume this should be done [here (link to code)](https://github.com/Rhyono/CraftStore/blob/master/CraftStore.lua#L1438).

## TL;DR

There's a need for a migration it seems.

**Comment by Zeratoxx** [comment]
It seems like you could painfully fix the issue I found with the command `/cspurge`.. but then you need to log in to each char again. And it would need big announcements for users.. not very practical I would say.

**Comment by Zeratoxx** [comment]
Nested PR initiated: https://github.com/Sc4rlite/CraftStore/pull/1

**Comment by Sc4rlite** [comment]
Personally I'm not sure if a rather large migration routine is the best approach to address one instance of a more widespread problem that has been occuring for a long time now.

If @Rhyono thinks otherwise feel free to chime in.

But if I got this correctly, your migration is because of wrong inventory caches, right? Something I noticed CraftStore to have problems with since I started using the addon. Wrongly cached items start to appear in the matrix on force quits, server timeouts or pc crashes and they never get cleaned up by the addon.

It seems instead of updating the inventory on login and bank open and deleting the wrongly cached zombie items from the cache, it seems CraftStore only reacts to item changes - add and remove. So an item that no longer exists can't ever be deleted from the matrix with regular means. A monthly /cspurge ended up a normal routine for my workflow. Gladly it doesn't delete other characters researched traits, not even the current timers.

@Zeratoxx my suggestion would be - instead of a 1-time migration routine - to solve the bug at its root and do a full upgrade to the inventory cache on login and bank open instead of an incremental one as it is currently. That would solve the above problem and all similar ones all at once. I'm afraid I don't have the expertise where or how that would need to be included but it seems you know your way around, if you are up to the task.

Again, if @Rhyono thinks your approach is sensible, I have no problem to merge it. Or maybe better if you open a separate pull request for your additions.

**Comment by Sc4rlite** [comment]
However, I noticed the same things that happened to jewelry tooltips and suggestions are also happening to nirnhoned shields, so I included a change to address that as well.
It seems an old incongruenty by ZOS needed shields to be handled differently than everythying else but that seems to be no longer needed anymore since Gold Road and to cause problems now.

**Comment by Zeratoxx** [comment]
> A monthly /cspurge ended up a normal routine for my workflow.

That genuinally surprises me, never needed and used this slash command 🤷 

> your migration is because of wrong inventory caches

Well.. the cache was correct in my experience before the indices got swapped by ZOS. The migration routine is looking for now faulty entries and "move them to the opposite side" to say so. That means that the migration routine expects jewelry entries at least.

> It seems instead of updating the inventory on login and bank open and deleting the wrongly cached zombie items from the cache, it seems CraftStore only reacts to item changes - add and remove.

Yes, as far as I understand the code, this is the expected behavior. There is a strong pro argument for this lean behavior: a full scan through the inventory costs computing time, doing this regularly will reduce the game performance drastically. But that's why `/cspurge` exists I assume.

> to solve the bug at its root

You already solved the bug at its root! Your change [here](https://github.com/Rhyono/CraftStore/pull/51/files#diff-1cfe4a1367862f6aed03b55f4cceb38f76856e9c788abfa977a9e6e45a3ff36eR3192) is the correct fix for the faulty behavior logicwise. But the cache wont get checked afterwards. That's fine as long as it get only updated with according entries which is the case normally.

> That would solve the above problem and all similar ones all at once

I never acknowledged any other issues, please enlighten me :)

> it seems you know your way around, if you are up to the task.

I just started reading through the code of CS the same day I created my PR at your repo, so I'm also not that aware I'm afraid. Additionally, with your suggestion "to solve the bug at its root and do a full upgrade to the inventory cache on login and bank open" you basically ask for an overhaul of the code -- which is as pretty big task imo.. I have neither the time nor the interest for that, in addition to my argument of the performance hit.

**Comment by Zeratoxx** [comment]
> I noticed the same things [...] are also happening to nirnhoned shields.

Possible, I can't check your changes at the moment. 🤷 :)

**Comment by Zeratoxx** [comment]
> I noticed the same things [...] are also happening to nirnhoned shields, so I included a change to address that as well.

Okay checked, solution works. This time there is no need for migration as far as I assess

**Review by Zeratoxx — APPROVED** [review]

**Comment by Rhyono** [comment]
It's been years since I've paid attention to the performance of addons, but if I assume it has not exponentially improved recently, then updating the entire cache upon login and/or bank access will cause a noticeable stutter/full freeze. This is why it uses the less foolproof transactional methodology and why cspurge exists. I wouldn't be against someone adding a setting to use an event trigger for doing full refreshes, but I never bothered to do it myself.

As for the migration: I can see the argument for and against it. Major issues (especially when introduced by ZOS) are reasonably expected to cause some level of inconvenience, but usually that inconvenience just trickles back down to the addon developers as users blame us/them. If it doesn't cause performance issues for anyone that is post migration, I don't mind it being included. Maybe at some point it should be undone, but end users do hate having to do anything themselves.

**Comment by Zeratoxx** [comment]
> If it doesn't cause performance issues for anyone that is post migration

@Rhyono the migration I wrote is a "run-once"-routine. I added a boolean saved var that saves if the routine got already executed.
The only code that adds up in `OnAddOnLoaded` is one if statement that checks the boolean flag and another if statement for debug.
I could remove the debug messages 🤷‍♂️.

File: CraftStore.lua [hunk 1]
```diff
@@ -3170,7 +3170,7 @@ function CS.GetTrait(link)
   elseif wt==WEAPONTYPE_HEALING_STAFF then
     craft=6;line=5;
   elseif wt==WEAPONTYPE_SHIELD then
-    craft=6;line=6;trait=trait-10;
+    craft=6;line=6;
```
Comment (Zeratoxx) [inline CraftStore.lua:3173]: Ah, This change is fixing nirnhoned, but breaks other trait like divines and so on..

Comment (Zeratoxx) [inline CraftStore.lua:3173]:
```suggestion
    craft=6;line=6;trait=trait-10;
```

Comment (Zeratoxx) [reply CraftStore.lua:3173]: And here https://github.com/Rhyono/CraftStore/pull/51/files#diff-1cfe4a1367862f6aed03b55f4cceb38f76856e9c788abfa977a9e6e45a3ff36eR3142: 
```
if at == ARMORTYPE_NONE then trait = 19 end
```

**Review by Zeratoxx — CHANGES_REQUESTED** [review]

File: CraftStore.lua (hunk 1 above)
Comment (Sc4rlite) [reply CraftStore.lua]: Good catch. I see now how it works.
On Live, I have no characters left to test other traits, and currently don't have PTS downloaded, so I can't test it myself. If it's no bother, please check if my commit matches your suggestions

**Comment by Sc4rlite** [comment]
Sounds good. I merged in the migration.
Like I mentioned, I don't understand enough of the ESO addon framework or Lua in general to grasp most of it and don't feel qualified to give a code review. If a review is wanted, I hope somebody else finds the time to look over it.
